### PR TITLE
delete chronicle data in the background

### PR DESCRIPTION
**Pull request by anzioka**

### CODE REVIEW

File: src/main/java/com/openlattice/chronicle/services/delete/DataDeletionService.java [hunk 1]
```diff
@@ -168,14 +171,16 @@ private void deleteStudyData(
...
+            executor.execute( () -> {
+                deleteParticipantNeighbors(
+                        dataApi,
+                        searchApi,
+                        srcEntitySetIds,
+                        dstEntitySetIds,
+                        participantsToDelete,
+                        participantsESID,
+                        deleteType );
+            } );
```
Comment (UnsungHero97) [inline src/main/java/com/openlattice/chronicle/services/delete/DataDeletionService.java:183]: `deleteParticipantNeighbors` calls `dataApi.deleteEntities` so it makes sense to wrap this. However, there is another call to `dataApi.deleteEntities` below on line 186 and 197, and I think that will also need to be moved to the background.

File: src/main/java/com/openlattice/chronicle/services/delete/DataDeletionService.java [hunk 2]
```diff
@@ -260,14 +265,16 @@ private void legacyDeleteStudyData(
...
+            executor.execute( () -> {
+                deleteParticipantNeighbors(
+                        chronicleDataApi,
+                        userSearchApi,
+                        srcEntitySetIds,
+                        dstEntitySetIds,
+                        participantsToDelete,
+                        participantsESID,
+                        deleteType );
+            } );
```
Comment (UnsungHero97) [inline src/main/java/com/openlattice/chronicle/services/delete/DataDeletionService.java:277]: Same comment about `dataApi.deleteEntities` below on line 280.

**Review by UnsungHero97 — CHANGES_REQUESTED** [review]

File: src/main/java/com/openlattice/chronicle/services/delete/DataDeletionService.java (hunk 1 above)
Comment (anzioka) [reply src/main/java/com/openlattice/chronicle/services/delete/DataDeletionService.java:183]: setting `blockUntilCompletion` to `true` in the DataApi.deleteEntities call executes in the background, no? In that case, I don't think there is need wrap this in a background thread

File: src/main/java/com/openlattice/chronicle/services/delete/DataDeletionService.java [hunk 3]
```diff
@@ -168,32 +171,36 @@ private void deleteStudyData(
...
-                    participantsESID,
-                    deleteType );
-
-            // delete participants
-            dataApi.deleteEntities( participantsESID, participantsToDelete, deleteType, false );
-            logger.info( "Deleting {} participants from study {} in org {}.",
-                    participantsToDelete.size(),
-                    studyId,
-                    organizationId );
+            executor.execute( () -> {
```
[the remark this answers is not in the document]
Comment (anzioka) [reply src/main/java/com/openlattice/chronicle/services/delete/DataDeletionService.java:174]: Possible approach to handle delete tasks:

1) User submits a delete request to delete participant data
2) Create a runnable for the heavy lifting stuff (filtered search to get entityKeyIds to delete, invoking the DataApi.deleteEntities). Serialize and write the runnable(s) to an entity set
3) Delete request returns and user is notified that the participant data has been scheduled for deletion

... A periodic task will fetch all saved runnables, deserialize and execute them in the background. On completion, remove entry 

If the server is ever restarted, uncompleted runnables will be "safe"

Thoughts on this approach? @UnsungHero97 @geekbeast

[the remark this answers is not in the document]
Comment (geekbeast) [reply src/main/java/com/openlattice/chronicle/services/delete/DataDeletionService.java]: What's the serialization strategy for the runnables? How do you guarantee that they can be resumed?

**Review by geekbeast — APPROVED** [review]
Shouldn't there be a delete type hard somewhere?

**Comment by anzioka** [comment]
> Shouldn't there be a delete type hard somewhere?

deleteType is a parameter in the endpoint method. Chronicle web calls the endpoint  with deleteType = Hard

**Review by UnsungHero97 — APPROVED** [review]

**Pull request merged**